### ISSUE TITLE
Wrong VPA is attempting to update pod resources and failing

### ISSUE THREAD

**Issue by WesCossick** [kubernetes/autoscaler#7499]
**Which component are you using?**:

vertical-pod-autoscaler

**What version of the component are you using?**:

Component version: Not sure; using the one auto-installed by GKE.

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: v1.30.5
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.5-gke.1699000
</pre></details>

**What environment is this in?**:

Google Cloud Platform's GKE.

**What did you expect to happen?**:

For the resources to be updated by the correct VPA.

**What happened instead?**:

The wrong VPA is updating pod resources and failing.

**How to reproduce it (as minimally and precisely as possible)**:

Create the following two VPAs:

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: example-1-vpa
spec:
  targetRef:
    apiVersion: apps/v1
    kind: CronJob
    name: example-1
  updatePolicy:
    updateMode: "Initial"
    minReplicas: 1
```

```yaml
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: example-2-vpa
spec:
  targetRef:
    apiVersion: apps/v1
    kind: CronJob
    name: example-2
  updatePolicy:
    updateMode: "Initial"
    minReplicas: 1
```

Create the following two CronJobs:

```yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: example-1
spec:
  schedule: * * * * *
  jobTemplate:
    spec:
      template:
        spec:
          containers:
            - name: example-1
              image: example:latest
```

```yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: example-2
spec:
  schedule: * * * * *
  jobTemplate:
    spec:
      template:
        spec:
          containers:
            - name: example-2
              image: example:latest
```

Using `describe pod`, see the following annotations on one of the cronjob's pods:

```
Annotations:          vpaObservedContainers: example-1
                      vpaUpdates: Pod resources updated by example-2-vpa: container 0: 
```

When this happens, the other cronjob's pods look correct and have their CPU and memory correctly set by the VPA:

```
Annotations:          vpaObservedContainers: example-2
                      vpaUpdates: Pod resources updated by example-2-vpa: container 0: cpu request, memory request
```

Which one is mixed up changes based on the order in which you create the VPAs it seems.

**Anything else we need to know?**:

I pared down the two `CronJob` resources to simplify them.

**Comment by raywainman** [comment]
I believe this issue would happen with the VPA deployed from OSS as well.

The way VPA knows which pods to target is that it will fetch the CronJob object:

https://github.com/bskiba/autoscaler/blob/8ff3b4fd47bcba514c9e904b6bee481ff0f3703e/vertical-pod-autoscaler/pkg/target/fetcher.go#L187-L192

And it fetches the `apiObj.Spec.JobTemplate.Spec.Template.Labels` value.

From your example above, you are not setting any labels in your `jobTemplate` which basically means it is matching all pods in the same namespace (I'd have to confirm in the code to be 100% sure but this is fairly likely).

So now the two VPAs are essentially racing and competing against each-other.

To fix this, you'll need to set some labels in your two CronJob templates to make sure you are differentiating them from each-other.

Hope that helps :)

**Comment by raywainman** [comment]
That's a great idea, it's an easy pitfall.

I thought we had something documenting this but I actually couldn't find anything.

Let me see if I can put together a PR with this.

Thanks @WesCossick!

**Comment by omerap12** [comment]
@raywainman I can make a PR for this, but are we sure it's not a bug? shouldn't we log a warning to let users know their VPA will target everything?

**Comment by adrianmoisey** [comment]
That reminds me... https://github.com/kubernetes/autoscaler/pull/6460 was made a long time ago that fixes this problem.
I haven't read the code, but may be it doesn't apply to CronJobs, or is bugged in some other way?

**Comment by omerap12** [comment]
I'm not sure if this qualifies as a bug, but the issue is that the function: https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/utils/vpa/api.go#L163 is designed to identify the *parent* of a pod. However, the behavior is different for CronJobs. In the case of CronJobs, they create Jobs, which in turn create Pods. As a result, the `ownerReference` of the Pod points to the Job, and the `ownerReference` of the Job points to the CronJob.

**Comment by adrianmoisey** [comment]
I wonder if GKE is running an old version of the VPA?

I tried using the latest OSS one, and I can't reproduce this behaviour. 

VPA [1.1.0](https://github.com/kubernetes/autoscaler/releases/tag/vertical-pod-autoscaler-1.1.0) has https://github.com/kubernetes/autoscaler/pull/6460 which fixes this bug.

**Comment by maxcao13** [comment]
I ran into this while using VPA 1.2.0 trying to provide recommendations for CronJobs and using `hack/vpa.up` in a kind cluster. These are my manifests: 

```yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "*/3 * * * *"
  jobTemplate:
    spec:
      template:
        metadata:
          labels:
            app: hello
        spec:
          containers:
          - name: hello
            image: busybox:1.28
            imagePullPolicy: IfNotPresent
            command:
            - /bin/sh
            - -c
            - |
              tries=0
              while [ $tries -lt 30 ]; do
                echo "Hello, World $tries!"
                tries=$((tries+1))
                sleep 6
              done
          restartPolicy: OnFailure
        
---
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: hello
spec:
  targetRef:
      apiVersion: "apps/v1"
      kind:       "CronJob"
      name:       "hello"
  updatePolicy:
    updateMode: "Initial"
    minReplicas: 1

```

I had to specify the Job labels like @raywainman said in order for it to start working :/. Seems like a bug/docs issue to me if reproducible.

**Comment by adrianmoisey** [comment]
You're right, thanks for clarifying that.

The VPA is definitely expecting a CronJob to have `spec.jobTemplate.spec.template.metadata.labels`, and fails silently when it doesn't.

**Comment by omerap12** [comment]
> jobTemplate

If that's the case, we should at least emit a log ( and maybe throw an error? )

**Comment by adrianmoisey** [comment]
> However, the [JobController even has a check for the case that two Jobs would use the same labelSelectors](https://github.com/kubernetes/kubernetes/blob/8a39b6062060ffb4fba52f53c011bd00428db194/pkg/controller/job/job_controller.go#L258-L274) and errors.
> So in the scenario that we're discussing here, I would expect that the Jobs were in an error state along the way?

I tried to reproduce this error, and I couldn't.
It seems that when a Job is created, if no labelSelector is set (as per the example above), then one is auto-generated.

It seems to be this part of code: https://github.com/kubernetes/kubernetes/blob/8a39b6062060ffb4fba52f53c011bd00428db194/pkg/registry/batch/job/strategy.go#L223-L226

Based on this, it seems that the CronJob <-> Job relationship in Kubernetes is sound (as you had mentioned with the OwnerRef). And the Job <-> Pod relationship is also sound, due to these labelSelectors being auto-generated. (It does seem that if a user sets `.spec.manualSelector` no their Job, they can override this behaviour and potentially cause harm)

This makes me assume that for the VPA to grab the labels of a CronJob the approach that Kubernetes is using, which makes me think that this is a bug.

**Comment by voelzmo** [comment]
Oooh, great find on the auto-generated labels and selectors on Job! There even is [good documentation](https://kubernetes.io/docs/concepts/workloads/controllers/job/#specifying-your-own-pod-selector) about this on the Job.

So that means from the CronJob docs perspective it makes sense to allow the user leaving this empty, but this doesn't work for VPA. If it _is_ empty, you will at least get a condition in the VPA status
```
status:
  conditions:
  - lastTransitionTime: "2024-12-11T10:56:03Z"
    message: No pods match this VPA object
    reason: NoPodsMatched
    status: "True"
    type: NoPodsMatched
```

And if then users configure the same labels on different CronJobs, this is _still_ fine from the CronJob and Job controller's perspective, it is just VPA requiring unique labels per CronJob. So this is definitely a VPA topic.

I'm still wondering why https://github.com/kubernetes/autoscaler/pull/6460 wouldn't help here? Shouldn't it detect the case where the labels match, but following the `ownerRef` doesn't match?

**Comment by adrianmoisey** [comment]
I think I've figured out what's going on.


https://github.com/bskiba/autoscaler/blob/35359b902ba00931fcad5be7c72a03bd7be39f7f/vertical-pod-autoscaler/pkg/target/fetcher.go#L167-L168

This piece of code grabs the labels from a CronJob's Job template, and assuming that the Pods will be discoverable by using those labels. Since there may not be any labels on the CronJob's spec, the labelSelector can be empty.

Later on, then `GetControllingVPAForPod()` is executed, it passes a list of VPAs + the discovered labelSelectors (in our case, an empty labelSelector) to `PodMatchesVPA()`: https://github.com/bskiba/autoscaler/blob/35359b902ba00931fcad5be7c72a03bd7be39f7f/vertical-pod-autoscaler/pkg/utils/vpa/api.go#L172-L175

This then uses those labelSelectors to find a match: https://github.com/bskiba/autoscaler/blob/35359b902ba00931fcad5be7c72a03bd7be39f7f/vertical-pod-autoscaler/pkg/utils/vpa/api.go#L99-L110

Since the labelSelector that came from the CronJob is empty, nothing is matched.


I think we could fix this if we found an active Job, and used its labelSelector to find Pods. I assume the implementation may end up being a little gross though.

**Comment by voelzmo** [comment]
still, after https://github.com/kubernetes/autoscaler/pull/6460, we also check the `ownerRef` to avoid cases where the labels are coincidentally matching – this should never have been the same and the original issue should not have occurred, right?

**Comment by adrianmoisey** [comment]
Not to my understanding. Since we're matching a Pod to the labels discovered by the VPA (which is empty), no match happens.

I think the purpose of https://github.com/kubernetes/autoscaler/pull/6460 is to filter out Pods that were matched, but shouldn't have been matched.

I think the CronJob situation is opposite, no pods are being matched, since the labelSelector can't be discovered.

**Comment by voelzmo** [comment]
Quoting from the OP:
>  Yep, that seems to be the issue! In the real environment where we were doing experimentation, the CronJob objects did actually have one label, but it was identical for each CronJob

They did have labels and they were the same on each CronJob. My initial reaction was that this should never have happened, but it might be that we're missing something from the original fix idea of checking the OwnerRef.

**Comment by adrianmoisey** [comment]
> They did have labels and they were the same on each CronJob. My initial reaction was that this should never have happened, but it might be that we're missing something from the original fix idea of checking the OwnerRef.

Can you reproduce this error? I can't.

**Comment by voelzmo** [comment]
I took some time today to reproduce and look at this, here are my findings:

If you _don't_ specify any labels on your Cronjob, your VPAs end up in `NoPodsMatched` (as @adrianmoisey documented above, VPA tries to fetch the selector, finds none and ends up not matching any pods). Therefore, recommender doesn't provide _any_ recommendation:

```
status:
  conditions:
  - lastTransitionTime: "2025-02-07T09:36:07Z"
    message: No pods match this VPA object
    reason: NoPodsMatched
    status: "True"
    type: NoPodsMatched
  - lastTransitionTime: "2025-02-07T09:36:07Z"
    message: No pods match this VPA object
    reason: NoPodsMatched
    status: "False"
    type: RecommendationProvided
  recommendation: {}
```

When you _do_ specify a label in the CronJob for `.spec.jobTemplate.spec.template.metadata.labels` _and_ use the same label for two Cronjobs, then everything still works as expected:

Pods for Job `hellohello` gets updated by VPA `hellohello` 
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    vpaObservedContainers: hellohello
    vpaUpdates: 'Pod resources updated by hellohello: container 0: '
  creationTimestamp: "2025-02-07T10:00:00Z"
  finalizers:
  - batch.kubernetes.io/job-tracking
  generateName: hellohello-28982040-
  labels:
    app: hellohello-world-job
    batch.kubernetes.io/controller-uid: 8a14f686-c06e-4acc-a4e7-f26364f3f1f3
    batch.kubernetes.io/job-name: hellohello-28982040
    controller-uid: 8a14f686-c06e-4acc-a4e7-f26364f3f1f3
    job-name: hellohello-28982040
  name: hellohello-28982040-vsrt9
  namespace: default
  ownerReferences:
  - apiVersion: batch/v1
    blockOwnerDeletion: true
    controller: true
    kind: Job
    name: hellohello-28982040
```

and Pods for Job `hellohello2` gets updated by VPA `hellohello2`:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    vpaObservedContainers: hellohello
    vpaUpdates: 'Pod resources updated by hellohello2: container 0: '
  creationTimestamp: "2025-02-07T10:00:00Z"
  finalizers:
  - batch.kubernetes.io/job-tracking
  generateName: hellohello2-28982040-
  labels:
    app: hellohello-world-job
    batch.kubernetes.io/controller-uid: 43bc2493-f228-4c6e-8a9f-195b899ad38c
    batch.kubernetes.io/job-name: hellohello2-28982040
    controller-uid: 43bc2493-f228-4c6e-8a9f-195b899ad38c
    job-name: hellohello2-28982040
  name: hellohello2-28982040-slkqf
  namespace: default
  ownerReferences:
  - apiVersion: batch/v1
    blockOwnerDeletion: true
    controller: true
    kind: Job
    name: hellohello2-28982040
```

This uses VPA from master and, as expected, the original issue is not reproducible. This leads me to believe that the version used by the OP was older than VPA 1.1.0 and did not contain the fix provided in https://github.com/kubernetes/autoscaler/pull/6460

TL;DR: I don't think we need to do anything here, and I would suggest to close this. WDYT?

**Comment by omerap12** [comment]
@voelzmo , thanks for deep dive into this! I agree that this should be documented. perhaps under [FAQ](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/docs/faq.md)?

**Comment by adrianmoisey** [comment]
> sorry, I was too quick here. The fact that `.spec.jobTemplate.spec.template.metadata.labels` is not allowed to be empty for CronJobs if you want them under VPA control is still true, so this needs to be documented.

Agreed on the documentation, and also this issue should stay open, as it is a bug
